### PR TITLE
Photon: Use correct dims for older images

### LINKED ISSUE #3087: Photon isn't working in 4.3.1 after latest update
I just upgraded to Jetpack  3.8.1 on my WordPress 4.3.1 sites, and noticed some issues with Photon. It's no longer resizing my images to the sizes defined in my themes. On one of my sites it's now loading in the full sized images (with the resize query). On another site that's using the Twenty Fifteen theme, it's loading in small 80x80 featured image sizes instead of 825x510.

### PR DESCRIPTION
**Pull request by kraftbj**
If you are outputting a post thumbnail at a custom size that did not exist when the image was uploaded (thus does not exist per WP), Photon in 3.8.1 would output it at full-size with the full-size values as the height/width on the img tag.

This PR uses a core function to confirm we only return a reasonable size.

You can test by adding a new image size, e.g.
`add_image_size('featured-thumbnail', 400, 400, true );`
and outputting `the_post_thumbnail( 'featured-thumbnail');` in your theme, within the loop.

Then attach an pre-existing image as the post thumbnail for a post, then view the post.

Expected: The returned image fit within 400x400, using Photon, with the current values on both the img tag and the resize string in the Photon URL.

Previously: The returned image was full size in both.

Core without Photon: Dims within the 400x400, but providing the full-size image relying on the browser respecting the img height/width.

cc: @dereksmart 

See #2918 
Fixes #3087, #3098

### CODE REVIEW

**Comment by dereksmart** [comment]
Tests well on 4.3.1 and 4.4

File: class.photon.php [hunk 1]
```diff
@@ -454,21 +454,25 @@ public function filter_image_downsize( $image, $attachment_id, $size ) {
...
-					$image_args = array(
-						'width'  => $image_meta['width'],
-						'height' => $image_meta['height'],
-						'crop'   => true
-					);
+					$image_args['width']  = $image_meta['width'];
+					$image_args['height'] = $image_meta['height'];
 				}
 
+				list( $image_args['width'], $image_args['height'] ) = image_constrain_size_for_editor( $image_args['width'], $image_args['height'], $size );
```
Comment (kraftbj) [inline class.photon.php]: Need to check http://php.net/manual/en/migration70.incompatible.php#migration70.incompatible.variable-handling.list.order

Comment (zinigor) [inline class.photon.php]: I think this only applies to assignments with literal array notation, doesn't it?

Comment (kraftbj) [inline class.photon.php]: I think that's right, but skittish. Haven't seen in Core/JP other uses of list to define directly values within an array. Another validation that we should unit test stuff :)

Comment (zinigor) [inline class.photon.php]: Unit test all the things :)

**Comment by ChrisBegley** [comment]
I just tested this in 4.3.1 and 4.4 and noticed a problem. Photon is using "fit" instead of "resize", so the cropping is off.

For example, my theme calls for a 340x220 cropped image:
`add_image_size( 'post-feature', 340, 220, true );`

With this new update, it's giving me 293x220:
http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/01/wx6cBg46010.jpg?fit=293,220

Instead of 340x220:
http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/01/wx6cBg46010.jpg?resize=340,220

**Comment by kraftbj** [comment]
@ChrisBegley -- What's the code that is loading that thumbnail so I can test? With the `true` declaration, it should be hard-cropping it....

Unless, the image isn't present in that small size still... https://github.com/Automattic/jetpack/pull/3089/files#diff-71f62265b9fb89148ee4f42b3c900faeR459 damn.

If you take out the new line 460, does it work as expected?

@dereksmart Since Photon already loads in the WP image sizes, including the crop value declared, should we ever actually manually declare the crop?

**Comment by ChrisBegley** [comment]
@kraftbj commenting out line 460 seemed to help a little bit. It uses "resize" now, but it's still pulling in the wrong dimensions. It's pulling in 293x220, but it should be 340x220:
http://i1.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/01/wx6cBg46010.jpg?resize=293,220

This is what's in my theme:
`<a href="<?php the_permalink(); ?>" rel="bookmark"><?php the_post_thumbnail( 'post-feature'); ?></a>`

**Comment by kraftbj** [comment]
@ChrisBegley I'm digging into it more, honing in on L474 (`image_constrain_size_for_editor`). Here's my thought process is that, when a local resized version of an image does not exist, we're seeing the image height/width to full size ( https://github.com/Automattic/jetpack/blob/fix/photon-new-thumbs/class.photon.php#L465-L472 ).

Then, we're running it through image_constrain_size_for_editor, which sends back dims for that full-size image based on what's best given the size that you're wanting. I haven't tested it to confirm, but it makes sense that it'd return dimensions proper to that image (with no notion of cropping).

We need to probably limit redeclaration of size only for full size images. Since "unknown" thumbnails will use the requested size (instead of full as it is at this moment), `image_constrain_size_for_editor` isn't needed as it is being used. We'll return images with the crop and size as determined by the `add_image_size` declaration and let Photon do the heavy lifting.

Then downside, then, is we're back to the issue of needing to know the dimensions at some point. We can't fully trust the size declaration since you can add an image size of "1200,0" (despite it being not suggested) or of "1200,9999".

In short, this is why the wontfix'd bug of not returning image dimensions has been left untouched for so many years. Not an easy problem. I'm going to think on this and get some more eyes to look at it.

**Comment by ChrisBegley** [comment]
@kraftbj thanks for looking into it. Some of that is over my head, but if it's something that can be fixed I'd be happy to help test it out when you're ready.

**Comment by kraftbj** [comment]
Thanks @ChrisBegley. Your testing and reporting have been vital for this effort.

**Comment by kraftbj** [comment]
I added some unit tests for a handful of options, including failing ones that I believe account for the cases reported. See message on 4119ddea2cb74666e33b922af39b4773c15f1c3a for full details.

**Comment by zinigor** [comment]
Other than some nitpicks this is a huge step forward, thank you very much for the effort!

**Comment by kraftbj** [comment]
@ChrisBegley I think we got it. Can you test this out and let us know if anything seems off. This time around, we added tests and ran it on all supported versions of WordPress to make sure we handled images (as I'd expect at least) in all of the various situations we could think up.

**Comment by ChrisBegley** [comment]
@kraftbj I just tested in 4.4 and it looks great! It took care of https://github.com/Automattic/jetpack/pull/3089#issuecomment-162250900

There's one small thing I noticed that may not even be a problem. My theme has this:
`add_image_size( 'post-hero', 800);`

Photon used to use the "w" query and return this:
http://i0.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/12/Joker_money.jpg?w=800

Now it's using "fit":
http://i0.wp.com/techyourworld.com/wordpress/wp-content/uploads/2014/12/Joker_money.jpg?fit=800%2C535

Chrome says that the "fit" image is 1 pixel off, and is actually 799x535 instead of 800x535. Not a huge deal, but just wanted to bring it up.

**Comment by kraftbj** [comment]
The `w` -> `fit` is expected. Basically, with the `w` we're telling Photon to give us something with that width but we don't know or care about the height. We do care now with the srcset changes, so we're using fit instead and passing both.

The 1px diff is interesting. It may be a method of rounding used one place but not the other. I'm leaning toward that not being a blocker to get this out to production since it does greatly make things better, and punting the 1px issue. @dereksmart thoughts?

**Comment by ChrisBegley** [comment]
I agree. Thanks for getting it fixed so quickly!

File: class.photon.php [hunk 2]
```diff
@@ -454,21 +454,26 @@ public function filter_image_downsize( $image, $attachment_id, $size ) {
 
 				$photon_args = array();
 
-				// `full` is a special case in WP
-				// To ensure filter receives consistent data regardless of requested size, `$image_args` is overridden with dimensions of original image.
-				if ( 'full' == $size || ! $image_meta = image_get_intermediate_size( $attachment_id, $size ) ) {
+				$image_meta = image_get_intermediate_size( $attachment_id, $size );
+
+				if ( 'full' != $size && ! $image_meta ) {
```
Comment (dereksmart) [inline class.photon.php]: we can simplify these conditionals if the `if ( 'full' == size )` is first, then `elseif ( ! image_meta )`
